### PR TITLE
(feat) O3-3901 Add support for program states in esm-patient-programs

### PR DESCRIPTION
**Pull request by ynurmahomed**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [x] My work includes tests or is validated by existing tests.

## Screenshots
Screenshot from 2024-09-03 09-49-36

### CODE REVIEW

File: packages/esm-patient-programs-app/src/programs/programs-form.workspace.tsx [hunk 1]
```diff
@@ -38,6 +39,7 @@ const createProgramsFormSchema = (t: TFunction) =>
     enrollmentDate: z.date(),
     completionDate: z.date().nullable(),
     enrollmentLocation: z.string(),
+    selectedState: z.string(),
   });
```
Comment (denniskigen) [inline packages/esm-patient-programs-app/src/programs/programs-form.workspace.tsx:44]: Then, we could fetch the Config schema here using the `useConfig` hook like so:

```tsx
const config = useConfig();
```

and then we can scope all the Program status logic via the boolean property we set earlier in the schema:

```tsx
config.showProgramStatusField && (
	style: { width: '50%' },
     legendText: '',
     value: stateSelect,
	}
)

etc...

**Review by denniskigen — COMMENTED** [review]
Great start, @ynurmahomed. Just a bunch of suggestions about using the config schema to make this new behaviour configurable.

File: packages/esm-patient-programs-app/src/programs/programs-detailed-summary.test.tsx [hunk 2]
```diff
@@ -68,6 +68,7 @@ describe('ProgramsDetailedSummary', () => {
     expect(screen.getByRole('columnheader', { name: /active programs/i })).toBeInTheDocument();
     expect(screen.getByRole('columnheader', { name: /date enrolled/i })).toBeInTheDocument();
     expect(screen.getByRole('columnheader', { name: /status/i })).toBeInTheDocument();
+    expect(screen.getByRole('columnheader', { name: /state/i })).toBeInTheDocument();
```
Comment (denniskigen) [inline packages/esm-patient-programs-app/src/programs/programs-detailed-summary.test.tsx:71]:
```suggestion
    expect(screen.getByRole('columnheader', { name: /program status/i })).toBeInTheDocument();
```

File: packages/esm-patient-programs-app/src/programs/programs-form.workspace.tsx [hunk 3]
```diff
@@ -38,6 +39,7 @@ const createProgramsFormSchema = (t: TFunction) =>
     enrollmentDate: z.date(),
     completionDate: z.date().nullable(),
     enrollmentLocation: z.string(),
+    selectedState: z.string(),
```
Comment (denniskigen) [inline packages/esm-patient-programs-app/src/programs/programs-form.workspace.tsx:42]:
```suggestion
    selectedProgramStatus: z.string(),
```

File: packages/esm-patient-programs-app/src/programs/programs-form.workspace.tsx [hunk 4]
```diff
@@ -262,6 +306,11 @@ const ProgramsForm: React.FC<ProgramsFormProps> = ({
       legendText: '',
       value: enrollmentLocation,
     },
+    {
+      style: { width: '50%' },
+      legendText: '',
+      value: stateSelect,
+    },
```
Comment (denniskigen) [inline packages/esm-patient-programs-app/src/programs/programs-form.workspace.tsx:313]:
```suggestion
    config.programStatus && {
      style: { width: '50%' },
      legendText: '',
      value: stateSelect,
    },
```

or something to that effect...

File: packages/esm-patient-programs-app/src/programs/programs.resource.tsx [hunk 5]
```diff
@@ -103,3 +103,12 @@ export const usePrograms = (patientUuid: string) => {
...
 };
+
+export const findLastState = (states: ProgramWorkflowState[]): ProgramWorkflowState => {
+  const active = states.filter((s) => !s.voided);
+  const notCompleted = active.find((s) => !s.endDate);
+  if (notCompleted) return notCompleted;
+  const compareFn = (a: ProgramWorkflowState, b: ProgramWorkflowState): number =>
+    new Date(a.endDate).getTime() - new Date(b.endDate).getTime();
+  return active.sort(compareFn).at(active.length - 1);
+};
```
Comment (denniskigen) [inline packages/esm-patient-programs-app/src/programs/programs.resource.tsx:114]: Does this work as well (I haven't tested it):

```suggestion
export const findLastState = (states: ProgramWorkflowState[]): ProgramWorkflowState => {
  const activeStates = states.filter(state => !state.voided);
  const ongoingState = activeStates.find(state => !state.endDate);
  
  if (ongoingState) {
    return ongoingState;
  }
  
  return activeStates.sort((a, b) => new Date(b.endDate).getTime() - new Date(a.endDate).getTime())[0];
};
```

Comment (ynurmahomed) [reply packages/esm-patient-programs-app/src/programs/programs.resource.tsx:114]: It should work and it is cleaner.

**Comment by ynurmahomed** [comment]
@denniskigen Added your suggestions, please review again.

**Review by denniskigen — APPROVED** [review]
LGTM. Thanks, @ynurmahomed.

**Pull request merged**